### PR TITLE
Autocomplete location

### PR DESCRIPTION
**Pull request by shulijones**
https://censusviz.appspot.com/
As the user types in the location box, offer autocompleted suggestions in addition to the hardcoded state options.

I've spent a good bit of time trying to get the autocomplete to merge nicely with the state dropdown, with mostly good results. One remaining issue is that the dropdown box does a little shake every time the autocomplete resets, which can be annoying. Also, I've got a delay on the autocomplete so that it doesn't try over and over as you type, and instead runs at intervals; getting the interval just right is tricky, so let me know if you think it's too long or too short.

### CODE REVIEW

File: src/main/webapp/query-control.js [hunk 1]
```diff
@@ -119,6 +121,7 @@ async function passQuery() {
     });
 }
 
+
```
Comment (kyracq) [inline src/main/webapp/query-control.js:124]: Can delete this extra line

File: src/main/webapp/query-control.js [hunk 2]
```diff
@@ -216,6 +224,75 @@ async function createStateDropdownList() {
...
+// Set up an autocomplete dropdown, attached to the main location input
+// and states dropdown, to suggest places to users as they type
+function setupAutocompleteLocation() {
+  const autocompleteInput = document.getElementById('location-list');
+  const autocompleteResults = document.getElementById('location');
+  const service = new google.maps.places.AutocompleteService();
+  const defaultLocationOptions = autocompleteResults.innerHTML; // all states
+
+  function displaySuggestions(predictions, status) {
+    if (status != google.maps.places.PlacesServiceStatus.OK) {
```
Comment (kyracq) [inline src/main/webapp/query-control.js:236]: Not sure how the status value works but could use a !== if possible

File: src/main/webapp/query-control.js [hunk 3]
```diff
@@ -216,6 +224,75 @@ async function createStateDropdownList() {
...
+      alert(
```
Comment (kyracq) [inline src/main/webapp/query-control.js:237]: When I start with the default location, click the dropdown, and click the default each U.S. state again, I get the zero results error. (Or if I just try typing "each "). Is there a way to not have the alert pop up for zero results and just show the dropdown with only states? Also, the input clears when there are no autocomplete results, though the input would be a business if it doesn't show in the suggestions, I'm wondering if users should be allowed to input anything?

File: src/main/webapp/query-control.js [hunk 4]
```diff
@@ -216,6 +224,75 @@ async function createStateDropdownList() {
...
+  // on the predictions it generated
+  autocompleteInput.addEventListener('input', debounce(function() {
+    const value = this.value;
+    value.replace('"', '\\"').replace(/^\s+|\s+$/g, ''); // Trim whitespace
+    service.getPlacePredictions({
+      'input': value,
+      'types': ['(regions)'], // no businesses, just cities counties etc.
+      'componentRestrictions': {'country': 'us'}, // USA only
+    }, displaySuggestions);
+  }, 250)); // wait 250 ms before sending the request so they don't overload
```
Comment (kyracq) [inline src/main/webapp/query-control.js:271]: I think a slightly shorter delay may be better, so users can see available suggestions before continuing typing. I also tried this without the delay locally and it seems to work fine with no delay, what issues did you run in to without the delay? I do notice that the shake becomes more obvious.

File: src/main/webapp/query-control.js (hunk 1 above)
Comment (shulijones) [reply src/main/webapp/query-control.js:124]: Done, thanks

File: src/main/webapp/query-control.js (hunk 2 above)
Comment (shulijones) [reply src/main/webapp/query-control.js:236]: Oh yeah probably, thanks

File: src/main/webapp/query-control.js (hunk 4 above)
Comment (shulijones) [reply src/main/webapp/query-control.js:271]: Yeah, I added the delay to try to get rid of the shake. Maybe I'm a bit sensitive but for me it makes it pretty hard to read the results list when it's at that level of shaking

Comment (shulijones) [reply src/main/webapp/query-control.js:271]: Tried out some options - 175 or 200 seem like they might be the sweet spot here for shaking vs too unresponsive.

File: src/main/webapp/query-control.js (hunk 3 above)
Comment (shulijones) [reply src/main/webapp/query-control.js:237]: Ah, that's a good point. I can just delete this if-clause and leave the next one that returns when there aren't any predictions. 
I'm not sure what you mean by your second question, sorry, could you elaborate?

Comment (kyracq) [reply src/main/webapp/query-control.js]: Oh after checking again the input comes back after you click away, as in if I type a non-location in, the zero results alert pops up, then I click ok, then the location input is blank, but then I click somewhere else and the non-location I typed comes back. Was wondering if the input can come back right after clicking OK

Comment (shulijones) [reply src/main/webapp/query-control.js]: Ah, gotcha. I think I deleted the alert entirely, so this shouldn't be a problem any more. I'll deploy the latest version now so you can take a look.

Comment (kyracq) [reply src/main/webapp/query-control.js]: Looks and works great now, thanks!

File: src/main/webapp/query-control.js [hunk 5]
```diff
@@ -216,6 +223,68 @@ async function createStateDropdownList() {
...
+  // on the predictions it generated
+  autocompleteInput.addEventListener('input', debounce(function() {
+    const value = this.value;
+    value.replace('"', '\\"').replace(/^\s+|\s+$/g, ''); // Trim whitespace
+    service.getPlacePredictions({
+      'input': value,
+      'types': ['(regions)'], // no businesses, just cities counties etc.
+      'componentRestrictions': {'country': 'us'}, // USA only
+    }, displaySuggestions);
+  }, 175)); // wait this many ms to send request so they don't overload
```
Comment (otobin) [inline src/main/webapp/query-control.js:263]: I didn't see earlier versions but I think 175 is good, I'm not seeing any shaking and I can both type fast and see results if I wait.

**Pull request merged**